README.md

```diff
@@ -22,14 +30,83 @@ dependencies: [
 ]
 ```
 
+## Usage
+
+`BaggageContext` is intended to be used in conjunction with the instrumentation of distributed systems. To make this
+instrumentation work, all parties involved operate on the same `BaggageContext` type. These are the three common
+parties, in no specific order, and guidance on how to use `BaggageContext`:
+
+### Frameworks/Libraries
+
+Developers creating frameworks/libraries (e.g. NIO, gRPC, AsyncHTTPClient, ...) which benefit from being instrumented
+should adopt `BaggageContext` as part of their public API. AsyncHTTPClient for example might accept a context like this:
+
+```swift
+let context = BaggageContext()
+client.get(url: "https://swift.org", context: context)
+```
+
+For more information on where to place this argument and how to name it, take a look at the
+["Context-Passing Guidelines"](#Context-Passing-Guidelines).
+
+Generally speaking, framework developers do not care about the specific contents of the `BaggageContext` but need to
+make sure to feed it through the entire system. At cross-cutting boundaries, e.g. right before sending an HTTP
+request, they'd inject the `BaggageContext` into the HTTP headers, allowing context propagation. On the flip-sight, an
+HTTP server should extract the request headers into a `BaggageContext`. Injecting/extracting is part of the
+`swift-tracing` libraries [and documented in its own repository](https://github.com/slashmo/gsoc-swift-tracing).
+
+### Instrumentation
+
+When implementing instrumentation for cross-cutting tools, `BaggageContext` becomes the way you propagate metadata such
+as trace ids. Because each instrument knows what values might be added to the `BaggageContext` they are the ones
+creating `BaggageContextKey` types dictating the type of value associated with each key added to the context:
+
+```swift
+private enum TraceIDKey: BaggageContextKey {
+  typealias Value = String
+}
+
+var context = BaggageContext()
+context[TraceIDKey.self] = "4bf92f3577b34da6a3ce929d0e0e4736"
+print(context[TraceIDKey.self] ?? "new trace id")
+```
+
+**TODO**: Add paragraph on why keys should be private/internal (#25)
+
+To make accessing values a bit more convenient, we encourage you to add computed properties to `BaggageContextProtocol`:
+
+```swift
+extension BaggageContextProtocol {
+  var traceID: String? {
+    get {
+      return self[TraceIDKey.self]
+    }
+    set {
+      self[TraceIDKey.self] = newValue
+    }
+  }
+}
+```
+
+### Distributed System
+
+When creating distributed systems, each of your components will likely make use of one or more generically instrumented
+frameworks/libraries. This means you get to choose what instrument(s) to use 🙌. Check out
+[the swift-tracing](https://github.com/slashmo/gsoc-swift-tracing) for instructions on how to get up & running.
```

Comment:
I'd not prefix this with "when creating distributed systems ..." at all, but simply say "you see some API that takes context, you pass whatever context you already have"

You can get inspiration on wording here: https://golang.org/pkg/context/

We should really focus on "you see a context? you pass one" if you don't have one, stop and think why, if you really dont have one, pass an empty one -- this relates to: https://github.com/slashmo/gsoc-swift-baggage-context/issues/26 as well

Reply:
Then after we explain "just always pass the context"

we can add a section on "This will enable you to enable tracing whenever your application grows", there will be no "oh no, now we have to fix all places to pass context" -- once you _suddenly_ need tracing because there's a fire on production or you're profiling some troublesome performance regressions, no-one will have the time to go through the entire system and start passing contexts around. You should always pass it, so that you're ready for when you need it -- your future self will be thankful for it 👍 

Or something like that